fastparquet/api.py

```diff
@@ -403,10 +415,150 @@ def remove_row_groups(self, rgs, write_fmd:bool = True,
+            By default, do not compress.
+            Please, review full description of this parameter in `write`
+            docstring.
+        write_fmd : bool, default True
+            Write updated common metadata to disk.
+        open_with : function
+            When called with a f(path, mode), returns an open file-like object.
+        mkdirs : function
+            When called with a path/URL, creates any necessary dictionaries to
+            make that location writable, e.g., ``os.makedirs``. This is not
+            necessary if using the simple file scheme.
+        rename : function
+            When called with a f(path1,path2), changes file path `path1` into
+            `path2`. Only used if `sort_pnames` is `True`.
+        stats : True|False|list of str
+            Whether to calculate and write summary statistics.
+            If True (default), do it for every column;
+            If False, never do;
+            If a list of str, do it only for those specified columns.
+        """
+        from .writer import write_simple, write_multi
+        partition_on = list(self.cats)
+        if isinstance(data, pd.DataFrame):
+            self_cols = sorted(self.columns + partition_on)
+            if self_cols != sorted(data.columns):
+                # TODO
+                # Proposal to clarify the legacy error message, by stating that
+                # column names of data on disk and column names of new data do
+                # not match, and showing inconsistent column indexes.
+                raise ValueError('File schema is not compatible with existing '
+                                 'file schema.')
+        if (self.file_scheme == 'simple'
+            or (self.file_scheme == 'empty' and self.fn[-9:] != '_metadata')):
+            # Case 'simple'.
+            write_simple(self.fn, data, self.fmd,
+                         row_group_offsets=row_group_offsets,
+                         compression=compression, open_with=open_with,
+                         has_nulls=None, append=True, stats=stats)
+        else:
+            # Case 'hive' or 'drill'.
+            write_multi(self.basepath, data, self.fmd,
+                        row_group_offsets=row_group_offsets,
+                        compression=compression, file_scheme=self.file_scheme,
+                        write_fmd=False, open_with=open_with, mkdirs=mkdirs,
+                        partition_on=partition_on, append=True, stats=stats)
+            if sort_key:
+                # Not using 'sort()' because 'row_groups' is a ThriftObject,
+                # not a list.
+                self.fmd.row_groups = sorted(self.fmd.row_groups, key=sort_key)
+            if sort_pnames:
+                self._sort_part_names(False, open_with, rename)
+            if write_fmd:
+                self._write_common_metadata(open_with)
+        self._set_attrs()
+
+    def _sort_part_names(self, write_fmd:bool=True, open_with=default_open,
```

Comment:
`_sort_part_names()` does modify metadata by modifying `file_path` when files are renamed.

Initially, I did call it with `write_fmd=True` from `ParquetFile.write_row_groups()`.
But then I realized that this function (`_sort_part_names()`) is not necessarily run every time by `ParquetFile.write_row_groups()` (this is user choice); So I could not rely on it to write `fmd` to disk.

I agree with your statement, hence the `_` in the name: `_sort_part_names()`.
I can move it out of `ParquetFile` class (removing the call to `_write_common_metadata`, but the signature will need to include more parameters from `ParquetFile`:
- `self.fs`
- `self.fmd.row_groups`
- `self.basepath`

I made the choice to keep it inside `ParquetFile` class in view of this.
Please, confirm me if I should move it out, I will do.

Reply:
OK, I see what you mean. This is getting to be an extremely long class. When we do eventually tidy up the code, it might be worth splitting into `ParquetFile(ParquetFileReader, ParquetFileModifier)` or something like that.